File: Algorithms/116. Populating Next Right Pointers in Each Node/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        if root is None or root.left is None:
            return
        root.left.next=root.right
        if root.next is not None:
            root.right.next=root.next.left
        self.connect(root.left)
        self.connect(root.right)

File: Algorithms/116. Populating Next Right Pointers in Each Node/Solution.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root!=null) {
            List<List<TreeLinkNode>> sumList=new ArrayList<List<TreeLinkNode>>();
            List<TreeLinkNode> initList=new ArrayList<TreeLinkNode>();
            initList.add(root);
            sumList.add(initList);
            for (int i = 0; i < sumList.size(); i++) {
                List<TreeLinkNode> list=sumList.get(i);
                List<TreeLinkNode> newList=new ArrayList<TreeLinkNode>();
                for (int j = 0; j < list.size(); j++) {
                    if (list.get(i).left!=null) {
                        newList.add(list.get(j).left);
                        newList.add(list.get(j).right);
                    }
                }
                if (newList.size()!=0) {
                    sumList.add(newList);
                }
            }
            for (int i = 0; i < sumList.size(); i++) {
                List<TreeLinkNode> list=sumList.get(i);
                for (int j = 0; j < list.size(); j++) {
                    if (j<list.size()-1) {
                        list.get(j).next=list.get(j+1);
                    }else {
                        list.get(j).next=null;
                    }
                }
            }
        }
    }
}